src/v/coproc/tests/fixtures/coproc_test_fixture.cc

```diff
@@ -82,103 +102,65 @@ ss::future<> coproc_test_fixture::restart() {
     co_await _root_fixture->wait_for_controller_leadership();
 }
 
-/// TODO: Code duplication remove after a rebase of dev
-static std::optional<kafka::partition_proxy> make_partition_proxy(
-  const model::materialized_ntp& ntp,
-  ss::lw_shared_ptr<cluster::partition> partition,
-  cluster::partition_manager& pm) {
-    if (!ntp.is_materialized()) {
-        return kafka::make_partition_proxy<kafka::replicated_partition>(
-          partition);
+ss::future<ss::stop_iteration> coproc_test_fixture::fetch_partition(
+  model::record_batch_reader::data_t& events,
+  model::offset& o,
+  model::topic_partition tp) {
+    auto response = co_await _client->fetch_partition(tp, o, 64_KiB, 1000ms);
+    if (response.data.error_code != kafka::error_code::none) {
+        co_return ss::stop_iteration::yes;
     }
-    if (auto log = pm.log(ntp.input_ntp()); log) {
-        return kafka::make_partition_proxy<kafka::materialized_partition>(*log);
+    vassert(response.data.topics.size() == 1, "Unexpected partition size");
+    auto& p = response.data.topics[0];
+    vassert(p.partitions.size() == 1, "Unexpected responses size");
+    auto& pr = p.partitions[0];
+    if (pr.error_code == kafka::error_code::none) {
+        auto crs = kafka::batch_reader(std::move(*pr.records));
+        while (!crs.empty()) {
+            auto kba = crs.consume_batch();
+            if (!kba.v2_format || !kba.valid_crc || !kba.batch) {
+                continue;
+            }
```

Comment:
You get some of this handling if you use the `record_batch_reader` interface that `crs` implements. The behaviour is slightly different, though, it will throw where you continue.

I'd be tempted to return `crs` here, and implement a consumer in the `drain` method below.

You may also want to skip control batches.

See an example here: https://github.com/graphcareful/redpanda/blob/c25156fdfc76a76de23adbf70184df1abb9e424c/src/v/pandaproxy/schema_registry/service.cc#L198-L203

Reply:
Great feedback! +1